# ATAC-seq
Analysis pipeline for ATAC-seq data

The following pipeline will describe the step-by-step analysis of ATAC-seq data (the **a**ssay for **t**ransposase-**a**ccessible **c**hromatin with **seq**uencing). This has been adapted from the following resources:

- https://vallierlab.wixsite.com/pipelines/atac-seq
- https://www.encodeproject.org/atac-seq/ *the recommended ENCODE pipeline*
- https://github.com/ENCODE-DCC/atac-seq-pipeline *the ENCODE pipeline available on github*
- https://github.com/harvardinformatics/ATAC-seq

An excellent recent review on the ATAC-seq analysis pipeline is reported by [(Yan et al. 2020)](https://genomebiology.biomedcentral.com/track/pdf/10.1186/s13059-020-1929-3).

The following steps will be covered:

- Sequencing quality control (QC) 
- Alignment 
- Peak Calling
- Visualisation
- Functional analysis & Motif Discovery


## Pre-alignment QC

- Sequence quality
- GC content
- Duplicate rate
- Length distribution 
- K-mer
- Adapter contamination

## Alignment

## Post-alignment QC

## Peak Calling 